### PR TITLE
Kaivos

### PR DESCRIPTION
**Pull request by kkmattil**
Kaivos manual.
1. Version

### CODE REVIEW

File: docs/data/kaivos/kaivos.md [hunk 1]
```diff
@@ -0,0 +1,30 @@
+# Kaivos User Guide #
+
+Relational databeses provide efifcient tool to manage structured data that keeps changing and whch contains data that needs to be accecced with complex queries. Client-server based SQL databases (e.g. MySQL, MariaDB, PostgreSQL) can not be effectively run in HPC computing enevioronments
```
Comment (amsaren) [inline docs/data/kaivos/kaivos.md]:
```suggestion
Relational databases provide efficient tool to manage structured data that keeps changing and which contains data that needs to be accessed with complex queries. Client-server based SQL databases (e.g. MySQL, MariaDB, PostgreSQL) can not be effectively run in HPC computing environments
```

File: docs/data/kaivos/kaivos_client_in_puhti.md [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+
+```text
+module load mysql
+```
+
+After this you can start the MySQL client program or execute some of the MySQL commands. A MySQL command line client session to kaivos.csc.fi is opened with the command:
+
+<pre>
+mysql -u <i>db_user_account</i> -p -h kaivos.csc.fi --local <i>db_name</i>
+</pre>
```
Comment (amsaren) [inline docs/data/kaivos/kaivos_client_in_puhti.md]: Formatting?

File: docs/data/kaivos/kaivos_client_in_puhti.md [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+
+This starts an interactive MySQL session which you can use to execute SQL commands for your database. In the command above options `-u` and `-h` define your database user account and the database server name (kaivos.csc.fi). The `-p` option defines that password is used for authentication. The `--local` option is not always necessary, but it is useful as it indicates to the MySQL client that the MySQL connection comes from a remote host.
+
+Instead of giving the SQL commands interactively, you can also write the SQL commands into a file and execute them with the command:
+
+<pre>mysql -u <i>db_user_account</i> -p -h kaivos.csc.fi --local <i>db_name</i> < <i>commands.sql</i> > <i>output.txt</i></pre>
```
Comment (amsaren) [inline docs/data/kaivos/kaivos_client_in_puhti.md]: Formatting?

File: docs/data/kaivos/kaivos_client_in_puhti.md [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+
+or
+<pre>mysql -u <i>db_user_account</i> -p -h kaivos.csc.fi --local \
+--skip-column-names --quick <i>db_name</i> < <i>commands.sql</i> > <i>output.txt</i></pre>
```
Comment (amsaren) [inline docs/data/kaivos/kaivos_client_in_puhti.md]: Formatting?

File: docs/data/kaivos/kaivos_client_in_puhti.md [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+You can now use SQL command `SHOW TABLES` to see which tables your database contains.
+<pre>
+mysql> <b>SHOW TABLES;</b>
++----------------+
+| Tables_in_DB_A |
++----------------+
+| results        |
++----------------+
+1 row in set (0.01 sec)
+</pre>
```
Comment (amsaren) [inline docs/data/kaivos/kaivos_client_in_puhti.md]: Formatting?

File: docs/data/kaivos/kaivos_client_in_puhti.md [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+mysql> INSERT INTO results (id, value, comment) VALUES (3, 25.33, "Value2");
+```
+When the table contains data, we can now do SQL queries with the SELECT command:
+<pre>mysql> <b>SELECT value FROM results WHERE id=2;</b>
++-------+
+| value |
++-------+
+| 12.33 |
++-------+
+</pre>
```
Comment (amsaren) [inline docs/data/kaivos/kaivos_client_in_puhti.md]: Formatting?

File: docs/data/kaivos/kaivos_import.md [hunk 7]
```diff
@@ -0,0 +1,54 @@
+# 3 Data import and export
+
+In the CSC computing environment we recommend command `mysqlimport` for loading datasets into the database.  This command reads in a delimited text file into a table already existing in the database. To load a large text file from a CSC computing server to a database in Kaivos, you can use the command syntax:
+<pre>
+mysqlimport -h kaivos.csc.fi -u <i>db_user_account</i> --local --compress --password <i>database_name</i> <i>input_file.table</i>
+</pre>
```
Comment (amsaren) [inline docs/data/kaivos/kaivos_import.md]: Formatting?

File: docs/data/kaivos/kaivos_import.md [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+
+Downloading complete tables or databases from the database can be done with command `mysqldump`. This command was developed for making backup copies of MySQL databases. In the case of kaivos.csc.fi, backup copies are not needed as the database is automatically backup copied by CSC.  Instead mysqldump offers an easy way to make a copy of your database so that you can move both the data content and structure of your database to another SQL server. You can make a copy of the whole database:
+
+<pre> mysqldump --set-gtid-purged=OFF -u <i>db_user_account</i> -h kaivos.csc.fi -p <i>database</i> > <i>database</i>_dump.txt </pre>
```
Comment (amsaren) [inline docs/data/kaivos/kaivos_import.md]: Formatting?

File: docs/data/kaivos/kaivos_import.md [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+
+or just from one or more tables:
+
+<pre> mysqldump --set-gtid-purged=OFF -u <i>db_user_account</i> -h kaivos.csc.fi -p <i>database</i> <i>table_name</i> > <i>table</i>_dump.txt</pre>
```
Comment (amsaren) [inline docs/data/kaivos/kaivos_import.md]: Formatting?

File: docs/data/kaivos/kaivos_import.md [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+
+
+When _mysqldump_ is used with the default settings, the result files contain MySQL commands that are needed to create and populate the selected database tables. Mysqldump locks the table in the beginning of the copying. Because of this only the _dbname_admin_ user account of the database can launch the command by default. In the case of other user accounts ( _dbname_read_ or _dbname_user_ )  `--skip-lock-tables` option should be added to the _mysqldump_ command.
+
+You can import the <i>database</i>_dump.txt file as follows:
+ 
+<pre>mysql -u <i>database_user_account</i> -h kaivos.csc.fi -p <i>database</i> < <i>database</i>_dump.txt</pre>
```
Comment (amsaren) [inline docs/data/kaivos/kaivos_import.md]: Formatting?

File: docs/data/kaivos/kaivos_import.md [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+<pre> mysqldump --set-gtid-purged=OFF -u <i>db_user_account</i> -h kaivos.csc.fi -p <i>database</i> > <i>database</i>_dump.txt </pre>
+
+or just from one or more tables:
+
+<pre> mysqldump --set-gtid-purged=OFF -u <i>db_user_account</i> -h kaivos.csc.fi -p <i>database</i> <i>table_name</i> > <i>table</i>_dump.txt</pre>
+
+
+When _mysqldump_ is used with the default settings, the result files contain MySQL commands that are needed to create and populate the selected database tables. Mysqldump locks the table in the beginning of the copying. Because of this only the _dbname_admin_ user account of the database can launch the command by default. In the case of other user accounts ( _dbname_read_ or _dbname_user_ )  `--skip-lock-tables` option should be added to the _mysqldump_ command.
+
+You can import the <i>database</i>_dump.txt file as follows:
```
Comment (amsaren) [inline docs/data/kaivos/kaivos_import.md]: Formatting?

File: docs/data/kaivos/kaivos_remote.md [hunk 12]
```diff
@@ -0,0 +1,25 @@
+# 5. Using kaivos.csc.fi directly from your local computer
+
+The MySQL databases in kaivos.csc.fi can be directly accessed only from the computing servers of CSC. However, you can make the database visible to your own computer using your CSC user account and port forwarding through an _ssh tunnel_.
+
+In linux and MacOSX machines an ssh tunnel from your local computer to kaivos.csc.fi via puhti.csc.fi can be done for example with the command:
+
+<pre>ssh -l <i>csc_username</i> -L 3306:kaivos.csc.fi:3306 puhti.csc.fi -N</pre>
```
Comment (amsaren) [inline docs/data/kaivos/kaivos_remote.md]: Formatting?

File: docs/data/kaivos/kaivos_remote.md [hunk 13]
```diff
@@ -0,0 +1,25 @@
...
+The `-N` option in the end of the connection command blocks the command shell after the connection is established. Once the connection is closed also the port forwarding becomes disabled. Note that the ssh command above uses your CSC user account and password, not the database user account.
+
+In Windows machines you can use e.g. plink program to open the tunnel. [Plink](https://www.chiark.greenend.org.uk/~sgtatham/putty/latest.html) can only be used through the command prompt. Below is a sample command for using plink:
+
+<pre>plink -L 3306:kaivos.csc.fi:3306 <i>taito_username</i>@puhti.csc.fi</pre>
```
Comment (amsaren) [inline docs/data/kaivos/kaivos_remote.md]: Formatting?

File: docs/data/kaivos/kaivos_remote.md [hunk 14]
```diff
@@ -0,0 +1,25 @@
...
+
+The tunneling commands above define that the communication to port 3306 (the default port of MySQL) in you local computer is transported to the MySQL port of _kaivos.csc.fi_ through _puhti.csc.fi_ server. As long as the connection to puhti.csc.fi is active the database in kaivos.csc.fi can be accessed from your local computer using the same MySQL commands as described above for Puhti (assuming that you have the MySQL client program installed in your local computer). The only difference compared to previous command examples is that the host section (-h) should now point to host `127.0.0.1` that refers to your local host, instead of kaivos.csc.fi.
+
+So for example the syntax to open an interactive MySQL session would now be:
+<pre>mysql -u <i>db_user_account</i> -p -h 127.0.0.1 --local <i>database_name</i></pre>
```
Comment (amsaren) [inline docs/data/kaivos/kaivos_remote.md]: Formatting?

File: docs/data/kaivos/kaivos_remote.md [hunk 15]
```diff
@@ -0,0 +1,25 @@
...
+
+And the syntax for mysqlimport:
+
+<pre>mysqlimport -h 127.0.0.1 -u <i>db_user_account</i> --local --compress --password <i>database_name</i> <i>input_file.table</i></pre>
```
Comment (amsaren) [inline docs/data/kaivos/kaivos_remote.md]: Formatting?

File: docs/data/kaivos/kaivos.md (hunk 1 above)
Comment (kkmattil) [reply docs/data/kaivos/kaivos.md]: OK

File: docs/data/kaivos/kaivos_client_in_puhti.md (hunk 2 above)
Comment (kkmattil) [reply docs/data/kaivos/kaivos_client_in_puhti.md]: OK

File: docs/data/kaivos/kaivos_client_in_puhti.md (hunk 3 above)
Comment (kkmattil) [reply docs/data/kaivos/kaivos_client_in_puhti.md]: OK

File: docs/data/kaivos/kaivos_client_in_puhti.md (hunk 4 above)
Comment (kkmattil) [reply docs/data/kaivos/kaivos_client_in_puhti.md]: OK

**Pull request merged**